### ISSUE TITLE
Chaining :hint and :follow-hint doesn't work with QtWebEngine

### ISSUE THREAD

**Issue by mbuelte** [qutebrowser/qutebrowser#3007]
I recently upgraded to qutebrowser version v0.11.0 and migrated my configuration to 'config.py', while setting up my bindings I noticed the following problem:

`config.bind('gi', 'hint inputs ;; follow-hint a', mode='normal')`

(I know, this is a little bit hacky, but 'a' is my first hintchar and for websites without many input fields this almost works. Since this version.)

Error message:
>   File "/usr/lib/python3.6/site-packages/qutebrowser/config/config.py", line 670, in early_init
>     raise configexc.ConfigFileErrors('config.py', config_api.errors)
> qutebrowser.config.configexc.ConfigFileErrors: Errors occurred while reading config.py:
>   While binding 'gi': follow-hint: This command is only allowed in hint mode, not normal.

qutebrowser version: v0.11.0-874-g248afde21 (2017-09-19 13:18:44 +0200)

**Comment by jgkamat** [comment]
This seems similar to #2989 (I think a fix for one should fix both)

**Comment by The-Compiler** [comment]
Seems like a duplicate of that.

**Comment by The-Compiler** [comment]
> also, it looks like 'follow-hint' is an alias you're using

[It isn't](
https://github.com/qutebrowser/qutebrowser/blob/master/doc/help/commands.asciidoc#follow-hint) :wink:

**Comment by mbuelte** [comment]
Great, thanks. I was looking for a similiar issue, but didn't find one.

**Comment by mbuelte** [comment]
I'm not pretty sure if https://github.com/qutebrowser/qutebrowser/commit/64b783d9c056045bcb5caa3d9ab8543b54fc4e18 fixed my issue with the above binding. I still get the error message "follow-hint: This command is only allowed in hint-mode, not normal.", when using the command. The only noticed change is, that this time qutebrowser came up even with this binding, instead of crashing directly on startup.

Running QB version: v0.11.0-1044-g0b5af757e (2017-09-28 21:51:07 +0200)

**Comment by The-Compiler** [comment]
The config actually isn't the problem here - a simple `:hint inputs ;; follow-hint a` never worked with QtWebEngine, even on v0.11.0.

**Comment by toofar** [comment]
This would probably be a pretty big job to get it working everywhere, a lot of commands invoke async Qt methods. Perhaps we should decide on a way (and syntax?) for passing around callbacks and start keeping track of the commands that people want this for the most and then do them first.

**Comment by The-Compiler** [comment]
@toofar Not sure if a callback is needed at all. Maybe something like a `command_finished` signal somewhere (and something in `@cmdutils.register()` to say that a command is expected to invoke it) would be enough, at least to make `;;` work. However, I'm not sure if it's worth the effort - especially for things like `:reload`, it'd get quite hairy as you'd need to connect that to some Qt signal and then disconnect it again.

Another possibility mentioned in IRC is using Python coroutines - while that'd require quite some code to interact with Qt properly, it might make a lot of code inside qutebrowser easier to read and thus be worth it anyways, see #3327. However, there's a lot of more important (and less intrusive) things to look at first.

**Comment by toofar** [comment]
To clarify, I'm not saying to use callbacks. I don't care how it is implmented. I'm saying that if making this happen for everything is too daunting (it is) then perhaps making it possible for some commands (like hint and reload) would be a good start. Discussing the downsides of a possiblle implementation instead a how we can find a way forward is not really helpful.

**Comment by The-Compiler** [comment]
I agree on that - there are likely many commands which don't need it at all (because they happen synchronously), so adding it gradually makes sense.

**Comment by The-Compiler** [comment]
FWIW it would probably also make sense to have some kind of "output" for commands, so that one could pipe commands into each other. IIRC Tridactyl has something like that. An example usecase: https://www.reddit.com/r/qutebrowser/comments/fq7euh/copyyank_jseval_output_to_clipboard/

**Comment by The-Compiler** [comment]
Another idea is having an exit status for commands and `&&` / `||` chainers - then you could do things like `:hint ... && spawn --userscript ...` to first have the user select an element, and then spawn a userscript on that element.

Not entirely sure if the possible use-cases are worth the added complexity, though.

**Comment by The-Compiler** [comment]
Another example for chaining: `:navigate next || scroll-to-perc` to either click a next link or scroll to the bottom.

**Comment by The-Compiler** [comment]
Yet [another idea](https://www.reddit.com/r/qutebrowser/comments/lzjmxk/given_an_url_focus_a_matching_tab_or_open_a_new/) around return values:

> Using `qutebrowser ":tab-select <pattern>"` does half of the job. But the shell return code for this command is always 0, no matter if matching tab is found or not, so I have no way to decide if I need to open a new tab. Is it doable in some other ways? Or, is it possible to introduce return codes into qutebrowser for such use-cases?

**Comment by The-Compiler** [comment]
And [another one](https://www.reddit.com/r/qutebrowser/comments/lzczkw/pop_up_a_dialog_from_userscript/) which comes up from time to time: Returning values to userscripts, e.g. by letting them run some kind of prompting command. Not sure yet if that's a good idea (compared to using the Python extension API), but it's worth a thought.

**Comment by The-Compiler** [comment]
Another one from #8021: Chaining `search` and `selection-follow`.
